core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,336 @@
+                if (cancelled) {
+                    upstream.cancel();
+                    return;
+                }
+
+                final Throwable cause = this.cause;
+                if (cause != null) {
+                    if (eventLoop.inEventLoop()) {
+                        onError0(cause);
+                    } else {
+                        eventLoop.execute(() -> onError0(cause));
+                    }
+                    return;
+                }
+
+                if (completing) {
+                    if (eventLoop.inEventLoop()) {
+                        onComplete0();
+                    } else {
+                        eventLoop.execute(this::onComplete0);
+                    }
+                    return;
+                }
+
+                if (demand() > 0) {
+                    // The 'demand' will be decreased by 'DefaultStreamMessage.notifySubscriberWithElements()'
+                    upstream.request(1);
+                }
+            }
+        }
+    }
+
+    @Override
+    void request(long n) {
+        if (initialized != 0 && demand() == 0) {
+            upstream.request(1);
+        }
+        super.request(n);
+    }
+
+    @Override
+    void cancel() {
+        cancelAndCleanup();
+        super.cancel();
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstreamUpdater.compareAndSet(this, null, subscription)) {
+            deferredInit(this.eventLoop);
+        } else {
+            subscription.cancel();
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject data) {
+        final EventExecutor eventLoop = this.eventLoop;
+        if (eventLoop.inEventLoop()) {
+            onNext0(data);
+        } else {
+            eventLoop.execute(() -> onNext0(data));
+        }
+    }
+
+    private void onNext0(HttpObject data) {
+        try {
+            process(data);
+
+            whenConsumed().thenRun(() -> {
+                if (demand() > 0) {
+                    // The `demand` will be decreased by 'DefaultStreamMessage.notifySubscriberWithElements()'
+                    upstream.request(1);
+                }
+            });
+        } catch (Throwable ex) {
+            handler.processOnError(ex);
+            cancelAndCleanup();
+            abort(ex);
+            Exceptions.throwIfFatal(ex);
+        }
+    }
+
+    private void cancelAndCleanup() {
+        cancelled = true;
+        final Subscription upstream = this.upstream;
+        if (upstream != null) {
+            upstream.cancel();
+        }
+        cleanup();
+    }
+
+    @Override
+    public void onError(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (cancelled) {
+            return;
+        }
+
+        final EventExecutor eventLoop = this.eventLoop;
+        if (eventLoop == null) {
+            this.cause = cause;
```

Comment:
Shouldn't we do this before `final EventExecutor eventLoop = this.eventLoop;` so that this cause is not swallowed?

Reply:
Oh. That's a good suggestion. 👍